### PR TITLE
Add list command

### PR DESCRIPTION
**Pull request by baduker**
List all Exercism tracks, exercises implementation, and download selected exercises.

**What does this PR do?**

The PR implements a feature in the exercism-cli for listing all the available Exercism tracks and exercises based on the given track and difficulty level. It also allows the users to download the list of exercises for a specific track and difficulty level.

**Changes Made:**

- Added a new command, `list`, to the CLI. This command can take the following flags `--track`, `--all`, `--level`, and `--download`.
- Implemented validation of the track names and difficulty levels using predefined maps and sets.
- Added a new feature to download all the exercises for a given track and difficulty level, provided the user is eligible for given exercises 
- Implemented sorting of exercises based on difficulty.
- Updated the CLI documentation to add the usage of the list command.
- Refactored `downlaod.go` command for readability
- Added new function `MakeRequest` to `api.go`

**Related Issue:**
- Resolves [#992](https://github.com/exercism/cli/issues/992)
- Resolves [#718](https://github.com/exercism/cli/issues/718)

**How to test it?**

- Check out the feature branch (on my fork): `add-list-command`
- Run `go build exercism/main.go`
- Run `./main configure --token=YOUR_TOKEN_VALUE`

Start testing the new command.

_Note: the results will be different because the CLI shows tracks and exercises available to a given user._

List your tracks: Run `./main list`

This should return:
```
+-----------------+--------+-----------+-----------+----------+--------------+---------------------------------------------+
| TRACK           | JOINED | EXERCISES | COMPLETED | PROGRESS | LAST TOUCHED | TRACK LINK                                  |
+-----------------+--------+-----------+-----------+----------+--------------+---------------------------------------------+
| AWK             | Yes    |        61 |         2 | 3%       | 2023-09-19   | https://exercism.org/tracks/awk             |
| Bash            | Yes    |        89 |         5 | 6%       | 2023-03-17   | https://exercism.org/tracks/bash            |
| C               | Yes    |        73 |         2 | 3%       | 2023-09-15   | https://exercism.org/tracks/c               |
| C#              | Yes    |       164 |         0 | 0%       | 2023-09-12   | https://exercism.org/tracks/csharp          |
| C++             | Yes    |        73 |         0 | 0%       | 2023-09-12   | https://exercism.org/tracks/cpp             |
| Erlang          | Yes    |        78 |         0 | 0%       | 2023-09-06   | https://exercism.org/tracks/erlang          |
| Go              | Yes    |       141 |        39 | 28%      | 2023-09-16   | https://exercism.org/tracks/go              |
| JavaScript      | Yes    |       141 |         3 | 2%       | 2023-07-28   | https://exercism.org/tracks/javascript      |
| Perl            | Yes    |        64 |         0 | 0%       | 2023-09-06   | https://exercism.org/tracks/perl5           |
| PL/SQL          | Yes    |        11 |         0 | 0%       | 2023-09-12   | https://exercism.org/tracks/plsql           |
| Python          | Yes    |       137 |        48 | 35%      | 2023-09-19   | https://exercism.org/tracks/python          |
| Rust            | Yes    |        97 |         7 | 7%       | 2023-09-14   | https://exercism.org/tracks/rust            |
| x86-64 Assembly | Yes    |        18 |         0 | 0%       | 2023-09-06   | https://exercism.org/tracks/x86-64-assembly |
+-----------------+--------+-----------+-----------+----------+--------------+---------------------------------------------+
| TOTAL TRACKS    | 13     |           |           |          |              |                                             |
+-----------------+--------+-----------+-----------+----------+--------------+---------------------------------------------+
```

**Note**: To see all tracks, run `./main list --all`.

List all exercises for a given track: Run `./main --list --track=awk`

This should return:

```
+-----------------------------+------------+-----------------------------------------------------------------------+
| EXERCISE                    | DIFFICULTY | EXERCISE LINK                                                         |
+-----------------------------+------------+-----------------------------------------------------------------------+
| High Scores                 | easy       | https://exercism.org/tracks/awk/exercises/high-scores                 |
| Allergies                   | easy       | https://exercism.org/tracks/awk/exercises/allergies                   |
| RNA Transcription           | easy       | https://exercism.org/tracks/awk/exercises/rna-transcription           |
| Leap                        | easy       | https://exercism.org/tracks/awk/exercises/leap                        |
| Scrabble Score              | easy       | https://exercism.org/tracks/awk/exercises/scrabble-score              |
| Hamming                     | easy       | https://exercism.org/tracks/awk/exercises/hamming                     |
| Acronym                     | easy       | https://exercism.org/tracks/awk/exercises/acronym                     |
| Two Fer                     | easy       | https://exercism.org/tracks/awk/exercises/two-fer                     |
| Pangram                     | easy       | https://exercism.org/tracks/awk/exercises/pangram                     |
| Grains                      | easy       | https://exercism.org/tracks/awk/exercises/grains                      |
| Series                      | easy       | https://exercism.org/tracks/awk/exercises/series                      |
| Resistor Color Duo          | easy       | https://exercism.org/tracks/awk/exercises/resistor-color-duo          |
| Largest Series Product      | easy       | https://exercism.org/tracks/awk/exercises/largest-series-product      |
| ETL                         | easy       | https://exercism.org/tracks/awk/exercises/etl                         |
| Rotational Cipher           | easy       | https://exercism.org/tracks/awk/exercises/rotational-cipher           |
| Queen Attack                | easy       | https://exercism.org/tracks/awk/exercises/queen-attack                |
| Matching Brackets           | easy       | https://exercism.org/tracks/awk/exercises/matching-brackets           |
| Triangle                    | easy       | https://exercism.org/tracks/awk/exercises/triangle                    |
| Armstrong Numbers           | easy       | https://exercism.org/tracks/awk/exercises/armstrong-numbers           |
| Protein Translation         | easy       | https://exercism.org/tracks/awk/exercises/protein-translation         |
| Bob                         | easy       | https://exercism.org/tracks/awk/exercises/bob                         |
| Space Age                   | easy       | https://exercism.org/tracks/awk/exercises/space-age                   |
| Secret Handshake            | easy       | https://exercism.org/tracks/awk/exercises/secret-handshake            |
| Darts                       | easy       | https://exercism.org/tracks/awk/exercises/darts                       |
| Gigasecond                  | easy       | https://exercism.org/tracks/awk/exercises/gigasecond                  |
| Sum of Multiples            | easy       | https://exercism.org/tracks/awk/exercises/sum-of-multiples            |
| Prime Factors               | easy       | https://exercism.org/tracks/awk/exercises/prime-factors               |
| Beer Song                   | easy       | https://exercism.org/tracks/awk/exercises/beer-song                   |
| Grade School                | easy       | https://exercism.org/tracks/awk/exercises/grade-school                |
| Robot Simulator             | easy       | https://exercism.org/tracks/awk/exercises/robot-simulator             |
| Word Count                  | easy       | https://exercism.org/tracks/awk/exercises/word-count                  |
| Resistor Color Trio         | easy       | https://exercism.org/tracks/awk/exercises/resistor-color-trio         |
| Raindrops                   | easy       | https://exercism.org/tracks/awk/exercises/raindrops                   |
| Hello World                 | easy       | https://exercism.org/tracks/awk/exercises/hello-world                 |
| Reverse String              | easy       | https://exercism.org/tracks/awk/exercises/reverse-string              |
| Rectangles                  | medium     | https://exercism.org/tracks/awk/exercises/rectangles                  |
| Pig Latin                   | medium     | https://exercism.org/tracks/awk/exercises/pig-latin                   |
| Two Bucket                  | medium     | https://exercism.org/tracks/awk/exercises/two-bucket                  |
| Affine Cipher               | medium     | https://exercism.org/tracks/awk/exercises/affine-cipher               |
| Wordy                       | medium     | https://exercism.org/tracks/awk/exercises/wordy                       |
| Anagram                     | medium     | https://exercism.org/tracks/awk/exercises/anagram                     |
| Kindergarten Garden         | medium     | https://exercism.org/tracks/awk/exercises/kindergarten-garden         |
| Luhn                        | medium     | https://exercism.org/tracks/awk/exercises/luhn                        |
| Tournament                  | medium     | https://exercism.org/tracks/awk/exercises/tournament                  |
| Matrix                      | medium     | https://exercism.org/tracks/awk/exercises/matrix                      |
| Run-Length Encoding         | medium     | https://exercism.org/tracks/awk/exercises/run-length-encoding         |
| Diamond                     | medium     | https://exercism.org/tracks/awk/exercises/diamond                     |
| OCR Numbers                 | medium     | https://exercism.org/tracks/awk/exercises/ocr-numbers                 |
| Atbash Cipher               | medium     | https://exercism.org/tracks/awk/exercises/atbash-cipher               |
| Roman Numerals              | medium     | https://exercism.org/tracks/awk/exercises/roman-numerals              |
| Forth                       | medium     | https://exercism.org/tracks/awk/exercises/forth                       |
| Grep                        | medium     | https://exercism.org/tracks/awk/exercises/grep                        |
| Nth Prime                   | medium     | https://exercism.org/tracks/awk/exercises/nth-prime                   |
| All Your Base               | medium     | https://exercism.org/tracks/awk/exercises/all-your-base               |
| Variable Length Quantity    | medium     | https://exercism.org/tracks/awk/exercises/variable-length-quantity    |
| Say                         | medium     | https://exercism.org/tracks/awk/exercises/say                         |
| Automated readability index | medium     | https://exercism.org/tracks/awk/exercises/automated-readability-index |
| Knapsack                    | hard       | https://exercism.org/tracks/awk/exercises/knapsack                    |
| List Ops                    | hard       | https://exercism.org/tracks/awk/exercises/list-ops                    |
| Change                      | hard       | https://exercism.org/tracks/awk/exercises/change                      |
| Mazy Mice                   | hard       | https://exercism.org/tracks/awk/exercises/mazy-mice                   |
+-----------------------------+------------+-----------------------------------------------------------------------+
| TOTAL EXERCISES:            | 61         |                                                                       |
+-----------------------------+------------+-----------------------------------------------------------------------+
```

Filter exercises for a given track using difficulty: Run `./main list --track=awk --level=hard`.

This should output:

```
+------------------+------------+-----------------------------------------------------+
| EXERCISE         | DIFFICULTY | EXERCISE LINK                                       |
+------------------+------------+-----------------------------------------------------+
| Knapsack         | hard       | https://exercism.org/tracks/awk/exercises/knapsack  |
| List Ops         | hard       | https://exercism.org/tracks/awk/exercises/list-ops  |
| Change           | hard       | https://exercism.org/tracks/awk/exercises/change    |
| Mazy Mice        | hard       | https://exercism.org/tracks/awk/exercises/mazy-mice |
+------------------+------------+-----------------------------------------------------+
| TOTAL EXERCISES: | 4          |                                                     |
+------------------+------------+-----------------------------------------------------+
```

Download selected exercises for a given track: Run `./main list --track=awk --level=hard --download`.

This should output:

```
Do you want to download 4 exercises? (y/n) y
Collecting exercises...
Downloading 27 files:  100% |████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| (27/27, 11 it/s)        

Successfully downloaded 27 files to the following directories:
/Users/baduker/Main/awk/knapsack
/Users/baduker/Main/awk/list-ops
/Users/baduker/Main/awk/change
/Users/baduker/Main/awk/mazy-mice
Happy coding!
```

### CODE REVIEW

File: cmd/configure.go [hunk 1]
```diff
@@ -116,7 +124,7 @@ func runConfigure(configuration config.Config, flags *pflag.FlagSet) error {
 
 	// Verify that the token is valid.
 	if !skipVerification {
-		client, err := api.NewClient(token, baseURL)
+		client, err := api.NewClient("", baseURL)
```
Comment (iHiD) [inline cmd/configure.go:127]: What does this do?

File: cmd/download_test.go [hunk 2]
```diff
@@ -25,9 +25,7 @@ func TestDownloadWithoutToken(t *testing.T) {
 
 	err := runDownload(cfg, pflag.NewFlagSet("fake", pflag.PanicOnError), []string{})
 	if assert.Error(t, err) {
-		assert.Regexp(t, "Welcome to Exercism", err.Error())
-		// It uses the default base API url to infer the host
-		assert.Regexp(t, "exercism.io/my/settings", err.Error())
+		assert.Regexp(t, "there is no token configured", err.Error())
```
Comment (iHiD) [inline cmd/download_test.go:28]: Why does this change?

File: cmd/download_test.go [hunk 3]
```diff
@@ -40,7 +38,7 @@ func TestDownloadWithoutWorkspace(t *testing.T) {
 
 	err := runDownload(cfg, pflag.NewFlagSet("fake", pflag.PanicOnError), []string{})
 	if assert.Error(t, err) {
-		assert.Regexp(t, "re-run the configure", err.Error())
+		assert.Regexp(t, "need an --exercise name or a solution --uuid", err.Error())
```
Comment (iHiD) [inline cmd/download_test.go:41]: Why does this change?

File: cmd/download_test.go [hunk 4]
```diff
@@ -54,7 +52,7 @@ func TestDownloadWithoutBaseURL(t *testing.T) {
...
```
Comment (iHiD) [inline cmd/download_test.go:55]: Why does this change?

File: cmd/download_test.go [hunk 5]
```diff
@@ -116,22 +114,13 @@ func TestSolutionFile(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 			sf := solutionFile{
-				path:    tc.file,
-				baseURL: "http://www.example.com/",
+				path:      tc.file,
+				sourceURL: "http://www.example.com/",
```
Comment (iHiD) [inline cmd/download_test.go:118]: Why does this change?

File: cmd/download_test.go [hunk 6]
```diff
@@ -116,22 +114,13 @@ func TestSolutionFile(t *testing.T) {
...
 			}
-
-			url, err := sf.url()
-			if err != nil {
-				t.Fatal(err)
-			}
-
-			if url != tc.expectedURL {
-				t.Fatalf("Expected URL '%s', got '%s'", tc.expectedURL, url)
-			}
```
Comment (iHiD) [inline cmd/download_test.go:134]: Why remove this block?

**Review by iHiD — CHANGES_REQUESTED** [review]
Thanks for your work on this. I've reopened the PR :)

I have three general comments:
1. I really like the list functionality. I'm :+1: on adding that.
2. I don't love the download all functionality, as that has the downside of marking all those exercises in progress. I'd like to have a wider community discussion about that first, and also take time to consider it from a product perspective.
3. There are quite a few other changes in this PR that make it hard to see exactly what's going on where. I've added questions to those for clarification.

Action suggestions:
1. I think there are three seperate PRs here.
a) The refactorings
b) The addition of List
c) The addition of Download

I'd like to see a PR just for (b) and potentially one for (a) to start with. Those will be more manageable to review. In its current state, there's so much happening in this PR, I think it will take a long time to get a proper review. I'd like to get the List stuff reviewed properly instead and be ale to deploy that, without waiting on the rest to get reviewesd.

The List functionality also needs a comprehensive test adding.

Thanks for your work on this. I hope that makes sense!

File: cmd/download_test.go (hunk 2 above)
Comment (iHiD) [reply cmd/download_test.go:28]: As far as I can see, all tests on `main` are green: https://github.com/exercism/cli/actions/runs/6098825660/job/16549554194

Comment (baduker) [reply cmd/download_test.go:28]: According to [CONTRIBUTING.MD](https://github.com/exercism/cli/blob/main/CONTRIBUTING.md) you can run tests locally with `go test ./...` and this fails on a clean clone and `master`. I get this:

```
pwd && go test ./...
/Users/baduker/Downloads/cli
?   	github.com/exercism/cli/browser	[no test files]
?   	github.com/exercism/cli/exercism	[no test files]
ok  	github.com/exercism/cli/api	0.254s
ok  	github.com/exercism/cli/cli	0.315s
--- FAIL: TestSubmitFiles (0.01s)
        Error Trace:    submit_test.go:275
        Error:		Not equal: "This is file 2." (expected)
        	        != "" (actual)
        
--- FAIL: TestSubmitFilesForTeamExercise (0.00s)
        Error Trace:    submit_test.go:471
        Error:		Not equal: "This is file 2." (expected)
        	        != "" (actual)
        
FAIL
FAIL	github.com/exercism/cli/cmd	0.529s
ok  	github.com/exercism/cli/config	0.370s
ok  	github.com/exercism/cli/debug	0.424s
ok  	github.com/exercism/cli/workspace	(cached)
FAIL
```

Comment (iHiD) [reply cmd/download_test.go:28]: Do you mean `master` or `main`? The primary branch on this repo is `main`.

Comment (baduker) [reply cmd/download_test.go:28]: Yes, I meant `main`, not `master`.

Comment (iHiD) [reply cmd/download_test.go:28]: Hmmm, I don't know why that's not working on your machine when it is working on our CI. Maybe a different version of Go? CCing @junedev, @andrerfcsantos, @kytrinyx and @ErikSchierboom to see if they have any ideas :)

Comment (nywilken) [reply cmd/download_test.go:28]: @iHiD @ErikSchierboom the failing tests happen when using a version of Go 1.17 and above. There was a change to the [mime/multipart pkg ](https://go.dev/doc/go1.17#mime/multipart) that changes the file path information being returned any files included in the request. It looks like there's an open PR for this fix https://github.com/exercism/cli/pull/1066 that can be merged in.

Comment (iHiD) [reply cmd/download_test.go:28]: I've replied there. I'd appreciate eyes on that too pls :)

https://github.com/exercism/cli/pull/1066/files#r1331620696

**Comment by baduker** [comment]
I'm closing the PR as this is going to be broken down into two (or more) PRs, starting with [this one here](https://github.com/exercism/cli/pull/1117).

**Pull request closed without merging**